torch/_export/passes/add_runtime_assertions_for_constraints_pass.py

```diff
@@ -0,0 +1,197 @@
+
+
+class AddRuntimeAssertionsForConstraintsPass(ExportPassBase):
+    def __init__(self) -> None:
+        super().__init__()
+        self.current_gm: Optional[torch.fx.GraphModule] = None
+
+    def _process_constraints(self, constraints) -> Dict[int, List[ConstraintSpec]]:
+        constraints_id_to_constraint: Dict[int, List[ConstraintSpec]] = defaultdict(
+            list
+        )
+        if constraints is None:
+            return constraints_id_to_constraint
+
+        # Convert simple sympy Integers into concrete int to
+        # insert into graph
+        def _convert_to_int(val):
+            if val == sympy.oo:
+                return math.inf
+            if isinstance(val, sympy.Integer):
+                return int(val)
+            raise RuntimeError(
+                "Export constraints cannot be non-integer expressions"
+            )
+
+        constraint_id_to_dim: Dict[int, Dict[int, List[Tuple[int, int]]]] = defaultdict(
+            lambda: defaultdict(list)
+        )
+        for constraint in constraints:
+            min_int_val = _convert_to_int(constraint["min"])
+            max_int_val = _convert_to_int(constraint["max"])
+            constraint_id_to_dim[constraint["t_id"]][constraint["dim"]].append(
+                (min_int_val, max_int_val)
+            )
+
+        # Merge the constraints into a single list of constraints
+        for t_id in constraint_id_to_dim:
+            for dim, constraints in constraint_id_to_dim[t_id].items():
+                min_vals = [x[0] for x in constraints]
+                max_vals = [x[1] for x in constraints]
+                min_val = sorted(min_vals, reverse=True)[0]
+                max_val = sorted(max_vals, reverse=False)[0]
+
+                assert min_val <= max_val
+
+                constraints_id_to_constraint[t_id].append(
+                    ConstraintSpec(constraint_dim=dim, min_val=min_val, max_val=max_val)
+                )
+
+        return constraints_id_to_constraint
+
+    def call(self, graph_module: torch.fx.GraphModule) -> PassResult:
+        # resets the counter
+        self.input_tracker = 0
+        self.current_gm = graph_module
+        assert isinstance(self.current_gm, torch.fx.GraphModule)
+        self.constraints = self._process_constraints(get_export_meta(self.current_gm).input_shape_constraints)
+        self.example_inputs = pytree.tree_flatten(get_export_meta(self.current_gm).example_inputs)[0]
+        return super().call(graph_module)
+
+    def _insert_specialized_shapes_assert(self, arg, dims, current_inp):
+        # we don't want to get shapes from meta as they will be symbolic
+        shapes = current_inp.shape
+        for dim in dims:
+            assert_msg = (
+                f"Input #{self.input_tracker}'s dimension #{dim} size is "
+                f"specialized at {shapes[dim]}"
+            )
+            dim_node = self.call_operator(
```

Comment:
Non blocking. Using call_operator here seems to make implementation of inline constraints more complicated: in the overriding call_operator function, we need to differentiate whether the call operator is from the original graph module or the newly inserted assertions. Can we use _fx to make it clear that these are the newly added nodes?

Reply:
Nvm, let's just test whether meta is empty or not to differentiate.